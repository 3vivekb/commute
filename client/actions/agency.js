import fetchAction from '@conveyal/woonerf/fetch'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addLocally = createAction('add agency')
export const createAgency = (newAgency) => {
  newAgency.id = uuid.v4()
  newAgency.organizations = []
  return [
    addLocally(newAgency),
    push(`/agency/${newAgency.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete agency')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/agency/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAgency = (id) => [deleteLocally(id), push('/')]

export const loadAgencies = () => fetchAction({
  url: '/api/agency',
  next: (res, err) => {
    if (!err) {
      return setAllLocally(res)
    }
  }
})

export const updateAgency = (agency) => [
  setLocally(agency),
  push(`/agency/${agency.id}`)
]
/* const updateOnServer = (agency) =>
  serverAction({
    url: `/api/agency/${agency.id}`,
    options: {
      body: JSON.stringify(agency),
      method: 'put'
    }
  }) */  // TODO: update on server

export const setLocally = createAction('set agency')
export const setAllLocally = createAction('set agencies')
