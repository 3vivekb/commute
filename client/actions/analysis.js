import moment from 'moment'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addAnalysis = createAction('add analysis')
export const createAnalysis = (newAnalysis, organizationId) => {
  newAnalysis.id = uuid.v4()
  newAnalysis.lastRunDateTime = moment().unix()
  newAnalysis.trips = []
  return [
    addAnalysis({ analysis: newAnalysis, organizationId }),
    push(`/organizations/${organizationId}/analysis/${newAnalysis.id}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete analysis')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/analysis/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteAnalysis = (id, organzationId) => [
  deleteLocally({ analysisId: id, organzationId }),
  push(`/organizations/${organzationId}`)
]
