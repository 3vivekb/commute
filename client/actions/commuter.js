import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'
import uuid from 'uuid'

const addCommuter = createAction('add commuter')
export const createCommuter = (commuter, groupId) => {
  commuter.groupId = groupId
  commuter.id = uuid.v4()
  return [
    addCommuter(commuter),
    push(`/group/${groupId}`)
  ] // TODO save to server
}

const deleteLocally = createAction('delete commuter')
/* const deleteOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'delete'
    }
  }) */ // TODO delete on server
export const deleteCommuter = (commuterId, groupId) => [
  deleteLocally(commuterId),
  push(`/group/${groupId}`)
]

const updateLocally = createAction('update commuter')
/* const updateOnServer = (id) =>
  serverAction({
    url: `/api/commuter/${id}`,
    options: {
      method: 'update'
    }
  }) */ // TODO update on server
export const updateCommuter = (commuter, groupId) => [
  updateLocally(commuter),
  push(`/group/${groupId}`)
]
