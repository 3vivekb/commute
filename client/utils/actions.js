import fetchAction from '@conveyal/woonerf/fetch'
import qs from 'qs'
import {push} from 'react-router-redux'
import {createAction} from 'redux-actions'

/**
 * Make generic model actions that communicate with a rest server
 *
 * @param  {Object} cfg An object with the following parameters:
 * - commands: An object with the keys being commands and the corresponding cfg of each command
 * -- redirectionStrategy: The redirection strategy to use upon success.
 *  Can be one of following: 'toEntity', 'toParent', 'toHome', undefined (no redirection)
 *  Defaults are as follows:
 *    Collection GET, GET: null
 *    Collection POST, PUT: toEntity
 *    DELETE: toParent
 * - parentKey: Key in model of parent's id
 *   (must be provided if using toParent redirection strategy)
 * - parentName: Name of app route to parent view
 *   (must be provided if using toParent redirection strategy)
 * - pluralName: Plural name of the model
 * - singularName: Singular name of the model
 * @return {Object}     The actions
 */
const makeGenericModelActions = (cfg) => {
  const {commands, parentKey, parentName, pluralName, singularName} = cfg
  const actions = {}

  // make local set actions
  const addLocally = createAction(`add ${singularName}`)
  const deleteLocally = createAction(`delete ${singularName}`)
  const setLocally = createAction(`set ${singularName}`)
  const setAllLocally = createAction(`set ${pluralName}`)

  const baseEndpoint = `/api/${singularName}`

  const redirectionStrategies = {
    'toEntity': (entity) => {
      return push(`/${singularName}/${entity._id}`)
    },
    'toHome': () => {
      return push('/')
    },
    'toParent': (entity) => {
      return push(`/${parentName}/${entity[parentKey]}`)
    }
  }

  /**
   * Do a redirect if needed
   *
   * @param  {Object} cfg Paremters as follows:
   *   - actions: The actions to append to
   *   - endpointCfg: The endpoint config
   *   - defaultStrategy: The default redire strategy to use (default: undefined)
   *   - redirectArgs: Arguments to provide to the redirectionStrategy
   */
  const doRedirectIfNecessary = (cfg) => {
    const redirectionStrategy = (cfg.endpointCfg.redirectionStrategy !== undefined
      ? cfg.endpointCfg.redirectionStrategy
      : cfg.defaultStrategy)

    if (redirectionStrategy) {
      const redirect = redirectionStrategies[redirectionStrategy](cfg.redirectArgs)
      if (redirect) {
        cfg.actions.push(redirect)
      }
    }
  }

  if (commands['Collection GET']) {
    actions.loadAll = (queryParams) => {
      // only include filteredKeys in querystring
      let queryString = ''
      if (queryParams) {
        queryString = qs.stringify(queryParams)
        if (queryString.length !== 0) {
          queryString = '?' + queryString
        }
      }

      // make request
      return fetchAction({
        next: (res, err) => {
          if (!err) {
            return setAllLocally(res.value)
          }
          // TODO handle error
        },
        url: baseEndpoint + queryString
      })
    }
  }

  if (commands['Collection POST']) {
    const endpointCfg = commands['Collection POST']
    actions.create = (newEntity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const createdEntity = res.value
            const actions = [
              addLocally(createdEntity)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toEntity',
              redirectArgs: createdEntity
            })
            return actions
          }
        },
        options: {
          body: newEntity,
          method: 'POST'
        },
        url: baseEndpoint
      })
    }
  }

  if (commands['DELETE']) {
    const endpointCfg = commands['DELETE']
    actions.delete = (entity) => {
      return fetchAction({
        next: (res, err) => {
          if (err) {
            // TODO handle error
          } else {
            const actions = [
              deleteLocally(entity._id)
            ]
            doRedirectIfNecessary({
              actions,
              endpointCfg,
              defaultStrategy: 'toParent',
              redirectArgs: entity
            })
            return actions
          }
        },
        options: {
          method: 'DELETE'
        },
        url: `${baseEndpoint}/${entity._id}`
      })
    }
  }

  if (commands['GET']) {
    actions.loadOne = (id) => fetchAction({
      next: (res, err) => {
        if (!err) {
          return setLocally(res.value)
        }
        // TODO handle error
      },
      url: `${baseEndpoint}/${id}`
    })
  }

  if (commands['PUT']) {
    const endpointCfg = commands['PUT']
    actions.update = (entity) => fetchAction({
      next: (res, err) => {
        if (err) {
          // TODO handle error
        } else {
          const updatedEntity = res.value
          const actions = [
            setLocally(updatedEntity)
          ]
          doRedirectIfNecessary({
            actions,
            endpointCfg,
            defaultStrategy: 'toEntity',
            redirectArgs: updatedEntity
          })
          return actions
        }
      },
      options: {
        body: entity,
        method: 'PUT'
      },
      url: `${baseEndpoint}/${entity._id}`
    })
  }

  return actions
}

export default makeGenericModelActions
