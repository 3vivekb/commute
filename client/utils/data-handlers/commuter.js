
export function loadDataIfNeeded (props, component) {
  const {commuter, loadCommuter, loadSite, params} = props

  // check to see if data needs to be loaded
  // for the create commuter page, the params will be empty, but the site needs to load
  // for the edit commuter page, we also need to load the commuter
  if (
    (!params || !params.commuterId || (params.commuterId && !commuter)) &&
    !component.loadingInitialData
  ) {
    // data needs to be loaded
    component.loadingInitialData = true
    loadSite({id: params.siteId})
    if (params.commuterId && !commuter) {
      // on edit commuter page and commuter hasn't loaded
      loadCommuter({id: params.commuterId})
    } else {
      // loading of the site on create page will not trigger a component update
      // therefore, say data has loaded
      component.setState({dataLoading: false})
    }
    return
  }
  component.loadingInitialData = false
  component.setState({dataLoading: false})
}
