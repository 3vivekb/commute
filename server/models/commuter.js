const Schema = require('mongoose').Schema

const geocodingPlugin = require('./plugins/geocode')
const trashPlugin = require('./plugins/trash')
const userPlugin = require('./plugins/user')

const modeType = {
  travelTime: Number
}

const schema = new Schema({
  modeStats: {
    bicycle: modeType,
    car: modeType,
    transit: modeType,
    walk: modeType,
    type: Schema.Types.Mixed
  },
  name: {
    required: true,
    type: String
  },
  siteId: {
    ref: 'Site',
    required: true,
    type: Schema.Types.ObjectId
  }
})

function postGeocodeHook (commuter) {
  console.log('commuter added or location changed, initiating polygon calculation')

  // import here to resolve circular import
  const models = require('./')
  const isochroneUtils = require('../utils/isochrones')

  models.Site.findOne({ _id: commuter.siteId, trashed: undefined, user: commuter.user })
    .exec()
    .then((site) => {
      if (site.calculationStatus === 'successfully') {
        models.Polygon.find({ _id: commuter.siteId })
          .exec()
          .then((polygons) => {
            const siteIsochrones = {}
            polygons.forEach((polygon) => {
              if (!siteIsochrones[polygon.mode]) {
                siteIsochrones[polygon.mode] = {
                  features: []
                }
              }
              siteIsochrones[polygon.mode].features.push(polygon)
            })

            if (site.calculationStatus === 'successfully') {
              isochroneUtils.calculateIsochroneStatsForCommuter(commuter, siteIsochrones)

              console.log('commuter stats calculated')

              commuter.save()
            }
          })
          .catch((err) => {
            console.error('error calculating commuter stats:', err)
          })
      }
    })
    .catch((err) => {
      console.error('error calculating commuter stats:', err)
    })
}

schema.plugin(geocodingPlugin(postGeocodeHook))
schema.plugin(trashPlugin)
schema.plugin(userPlugin)

module.exports = schema
