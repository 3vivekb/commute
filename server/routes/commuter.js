const models = require('../models')
const endpointFactory = require('../utils/restEndpoints')

module.exports = function makeRoutes (app, jwt) {
  const baseCfg = {
    model: models.Commuter,
    name: 'commuter'
  }
  endpointFactory.makeRestEndpoints(
    app,
    jwt,
    Object.assign(
      {
        commands: {
          'Collection GET': {},
          'Collection POST': {},
          GET: {},
          DELETE: {},
          PUT: {}
        }
      },
      baseCfg
    )
  )
  endpointFactory.makePublicRestEndpoints(app, baseCfg)
}
