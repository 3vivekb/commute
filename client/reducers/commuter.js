import {addToEntityMap} from '../utils/entities'
import {
  getGenericReducerInitialState,
  makeGenericReducerHandlers,
  updateTime
} from '../utils/reducers'

export const reducers = makeGenericReducerHandlers({
  handlers: ['add', 'add many', 'delete', 'set', 'set many'],
  name: {
    singular: 'commuter',
    plural: 'commuters'
  }
})

reducers['add group'] = function (state, action) {
  const newGroup = {...action.payload}
  let updatedState = {...state}
  newGroup.commuters.map((commuter) => {
    updatedState = addToEntityMap(updatedState, commuter)
  })
  return updateTime(updatedState)
}

reducers['append commuters'] = function (state, action) {
  let updatedState = state
  action.payload.commuters.forEach((commuter) => {
    updatedState = addToEntityMap(updatedState, commuter)
  })
  return updateTime(updatedState)
}

export const initialState = getGenericReducerInitialState()
