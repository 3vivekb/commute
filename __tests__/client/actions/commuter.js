/* global describe, it */

import {expectCreateCommuter, expectDeleteCommuter, expectUpdateAction} from '../../test-utils/actions'
import * as commuter from '../../../client/actions/commuter'

describe('actions > commuter', () => {
  it('create commuter should work', () => {
    const data = {
      name: 'mockCommuter'
    }
    const actions = commuter.createCommuter(data, 'group-id')

    expectCreateCommuter(actions)
  })

  it('update commuter should work', () => {
    const data = {
      id: 'commuter-id',
      name: 'mockCommuter'
    }
    const actions = commuter.updateCommuter(data)

    expectUpdateAction(actions)
  })

  it('delete commuter should work', () => {
    const result = commuter.deleteCommuter('commuter-id', 'group-id')

    expectDeleteCommuter(result)
  })
})
