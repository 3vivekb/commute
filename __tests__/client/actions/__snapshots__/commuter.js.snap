exports[`actions > commuter create commuter should work 1`] = `
Object {
  "payload": Object {
    "groupId": "group-id",
    "name": "mockCommuter",
  },
  "type": "add commuter",
}
`;

exports[`actions > commuter create commuter should work 2`] = `
Object {
  "payload": Object {
    "args": Array [
      "/group/group-id",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;

exports[`actions > commuter delete commuter should work 1`] = `
Array [
  Object {
    "payload": "commuter-id",
    "type": "delete commuter",
  },
  Object {
    "payload": Object {
      "args": Array [
        "/group/group-id",
      ],
      "method": "push",
    },
    "type": "@@router/CALL_HISTORY_METHOD",
  },
]
`;

exports[`actions > commuter update commuter should work 1`] = `
Object {
  "payload": Object {
    "id": "commuter-id",
    "name": "mockCommuter",
  },
  "type": "update commuter",
}
`;

exports[`actions > commuter update commuter should work 2`] = `
Object {
  "payload": Object {
    "args": Array [
      "/group/undefined",
    ],
    "method": "push",
  },
  "type": "@@router/CALL_HISTORY_METHOD",
}
`;
